rpc/rpc.go

```diff
@@ -1256,245 +1249,234 @@ func (c *Conn) handleRelease(ctx context.Context, id exportID, count uint32) err
 	}
 
 	switch d.Context().Which() {
 	case rpccp.Disembargo_context_Which_receiverLoopback:
+		defer release()
+
 		id := embargoID(d.Context().ReceiverLoopback())
 		c.mu.Lock()
 		e := c.findEmbargo(id)
 		if e == nil {
 			c.mu.Unlock()
-			return failedf("incoming disembargo: received sender loopback for unknown ID %d", id)
+			return rpcerr.Failedf("incoming disembargo: received sender loopback for unknown ID %d", id)
 		}
 		// TODO(soon): verify target matches the right import.
 		c.embargoes[id] = nil
 		c.embargoID.remove(uint32(id))
 		c.mu.Unlock()
 		e.lift()
+
 	case rpccp.Disembargo_context_Which_senderLoopback:
-		c.mu.Lock()
-		if tgt.which != rpccp.MessageTarget_Which_promisedAnswer {
-			c.mu.Unlock()
-			return failedf("incoming disembargo: sender loopback: target is not a promised answer")
-		}
-		ans := c.answers[tgt.promisedAnswer]
-		if ans == nil {
-			c.mu.Unlock()
-			return failedf("incoming disembargo: unknown answer ID %d", tgt.promisedAnswer)
-		}
-		if ans.flags&returnSent == 0 {
-			c.mu.Unlock()
-			return failedf("incoming disembargo: answer ID %d has not sent return", tgt.promisedAnswer)
-		}
-		if ans.err != nil {
-			c.mu.Unlock()
-			return failedf("incoming disembargo: answer ID %d returned exception", tgt.promisedAnswer)
-		}
-		content, err := ans.results.Content()
-		if err != nil {
-			c.mu.Unlock()
-			return failedf("incoming disembargo: read answer ID %d: %v", tgt.promisedAnswer, err)
-		}
-		ptr, err := capnp.Transform(content, tgt.transform)
+		var (
+			imp    *importClient
+			client *capnp.Client
+		)
+
+		syncutil.With(&c.mu, func() {
+			if tgt.which != rpccp.MessageTarget_Which_promisedAnswer {
+				err = rpcerr.Failedf("incoming disembargo: sender loopback: target is not a promised answer")
+				return
+			}
+
+			ans := c.answers[tgt.promisedAnswer]
+			if ans == nil {
+				err = rpcerr.Failedf("incoming disembargo: unknown answer ID %d", tgt.promisedAnswer)
+				return
+			}
+			if ans.flags&returnSent == 0 {
+				err = rpcerr.Failedf("incoming disembargo: answer ID %d has not sent return", tgt.promisedAnswer)
+				return
+			}
+
+			if ans.err != nil {
+				err = rpcerr.Failedf("incoming disembargo: answer ID %d returned exception", tgt.promisedAnswer)
+				return
+			}
+
+			var content capnp.Ptr
+			if content, err = ans.results.Content(); err != nil {
+				err = rpcerr.Failedf("incoming disembargo: read answer ID %d: %v", tgt.promisedAnswer, err)
+				return
+			}
+
+			ptr, err := capnp.Transform(content, tgt.transform)
+			if err != nil {
+				err = rpcerr.Failedf("incoming disembargo: read answer ID %d: %v", tgt.promisedAnswer, err)
+				return
+			}
+
+			iface := ptr.Interface()
+			if !iface.IsValid() || int64(iface.Capability()) >= int64(len(ans.resultCapTable)) {
+				err = rpcerr.Failedf("incoming disembargo: sender loopback requested on a capability that is not an import")
+				return
+			}
+
+			client := ans.resultCapTable[iface.Capability()].AddRef()
```

Comment:
I spotted a `LEAK: leaked client created by AddRef on /Users/lthibault/Go/src/capnproto.org/go/capnp/rpc/rpc.go:1333`, while messing around with benchmarks.  I ran the command `go test -bench=. ./...`, but the output doesn't make it clear exactly which test is triggering this.  I tried commenting out the `.AddRef()` part of the line and the error goes away without introducing any tests failures, but I'm not super confident.  Any thoughts?

Full output below:

```
 $ go test -bench=. ./...
goos: darwin
goarch: amd64
pkg: capnproto.org/go/capnp/v3
cpu: Intel(R) Core(TM) i7-1068NG7 CPU @ 2.30GHz
BenchmarkTextMovementBetweenSegments-8              3892            317708 ns/op
BenchmarkMarshal-8                               2416864               499.6 ns/op           264 B/op          3 allocs/op
BenchmarkUnmarshal-8                             2150120               555.0 ns/op           192 B/op          3 allocs/op
BenchmarkUnmarshal_Reuse-8                       2972167               419.1 ns/op             0 B/op          0 allocs/op
BenchmarkDecode-8                                    340           3256078 ns/op         294.83 MB/s     2880312 B/op      40002 allocs/op
BenchmarkDecode_Reuse-8                              727           1662096 ns/op         577.58 MB/s         400 B/op          3 allocs/op
BenchmarkGrowth_SingleSegment-8                      164           7588045 ns/op         138.19 MB/s     4343117 B/op         10 allocs/op
BenchmarkGrowth_MultiSegment-8                       100          10035792 ns/op         104.49 MB/s     1648598 B/op         21 allocs/op
BenchmarkSmallMessage_SingleSegment-8            1735057               689.2 ns/op       104.47 MB/s        1352 B/op          5 allocs/op
BenchmarkSmallMessage_MultiSegment-8             2053214               583.9 ns/op       123.32 MB/s        1216 B/op          4 allocs/op
PASS
ok      capnproto.org/go/capnp/v3       16.372s
PASS
ok      capnproto.org/go/capnp/v3/capnpc-go     0.444s
PASS
ok      capnproto.org/go/capnp/v3/encoding/text 0.120s
PASS
ok      capnproto.org/go/capnp/v3/exc   0.130s
PASS
ok      capnproto.org/go/capnp/v3/flowcontrol   0.154s
?       capnproto.org/go/capnp/v3/internal/aircraftlib  [no test files]
?       capnproto.org/go/capnp/v3/internal/capnptool    [no test files]
?       capnproto.org/go/capnp/v3/internal/chanmutex    [no test files]
PASS
ok      capnproto.org/go/capnp/v3/internal/demo 0.138s
?       capnproto.org/go/capnp/v3/internal/demo/books   [no test files]
PASS
ok      capnproto.org/go/capnp/v3/internal/mpsc 0.231s
?       capnproto.org/go/capnp/v3/internal/nodemap      [no test files]
goos: darwin
goarch: amd64
pkg: capnproto.org/go/capnp/v3/internal/packed
cpu: Intel(R) Core(TM) i7-1068NG7 CPU @ 2.30GHz
BenchmarkPack-8                   102496             11517 ns/op         889.12 MB/s
BenchmarkUnpack-8                 127020              9691 ns/op        1056.62 MB/s
BenchmarkUnpack_Large-8           784611              1401 ns/op        37386.49 MB/s
BenchmarkReader-8                  66501             17918 ns/op         571.48 MB/s
BenchmarkReader_Large-8            35362             34034 ns/op        1538.70 MB/s
PASS
ok      capnproto.org/go/capnp/v3/internal/packed       8.488s
?       capnproto.org/go/capnp/v3/internal/schema       [no test files]
?       capnproto.org/go/capnp/v3/internal/strquote     [no test files]
?       capnproto.org/go/capnp/v3/internal/syncutil     [no test files]
goos: darwin
goarch: amd64
pkg: capnproto.org/go/capnp/v3/pogs
cpu: Intel(R) Core(TM) i7-1068NG7 CPU @ 2.30GHz
BenchmarkExtract-8         67496             17730 ns/op           15184 B/op         32 allocs/op
BenchmarkInsert-8          67862             17781 ns/op           15164 B/op         29 allocs/op
PASS
ok      capnproto.org/go/capnp/v3/pogs  2.991s
LEAK: leaked client created by AddRef on /Users/lthibault/Go/src/capnproto.org/go/capnp/rpc/rpc.go:1333
goos: darwin
goarch: amd64
pkg: capnproto.org/go/capnp/v3/rpc
cpu: Intel(R) Core(TM) i7-1068NG7 CPU @ 2.30GHz
BenchmarkPingPong-8        44317             26844 ns/op            8888 B/op         96 allocs/op
PASS
exit status 1
FAIL    capnproto.org/go/capnp/v3/rpc   2.148s
?       capnproto.org/go/capnp/v3/rpc/internal/testcapnp        [no test files]
PASS
ok      capnproto.org/go/capnp/v3/schemas       0.117s
PASS
ok      capnproto.org/go/capnp/v3/server        0.123s
?       capnproto.org/go/capnp/v3/std/capnp/compat/json [no test files]
?       capnproto.org/go/capnp/v3/std/capnp/cxx [no test files]
?       capnproto.org/go/capnp/v3/std/capnp/persistent  [no test files]
?       capnproto.org/go/capnp/v3/std/capnp/rpc [no test files]
?       capnproto.org/go/capnp/v3/std/capnp/rpctwoparty [no test files]
?       capnproto.org/go/capnp/v3/std/capnp/schema      [no test files]
?       capnproto.org/go/capnp/v3/std/capnp/stream      [no test files]
?       capnproto.org/go/capnp/v3/std/go        [no test files]
FAIL
```

Reply:
Hypothesis: this should be `client =` rather than `client :=`; this is a new variable that is shadowing the old one, and thus the call to Release() on the old variable ends up not doing what we want.